File: frontend/types/prediction.ts
export interface ModelInfo {
  name: string;
  accuracy: number;
  roc_auc: number;
  threshold: number;
}

export interface AvailableModelsResponse {
  models: ModelInfo[];
  default_model: string;
}

export interface DiseaseInfo {
  id: string;
  label: string;
  nci_id: number;
}

export interface AvailableDiseasesResponse {
  diseases: DiseaseInfo[];
  default_disease: string;
}

export interface PredictionResult {
  smiles: string;
  model_name: string;
  prediction: "Active" | "Inactive";
  prediction_label: number;
  probability: number;
  threshold: number;
}

export interface TopSubstructure {
  token: string;
  description: string;
  score: number;
  percentage: number;
  occurrences: number;
  direction: "supporting" | "opposing";
}

export interface HeatmapResult {
  smiles: string;
  model_name: string;
  prediction: "Active" | "Inactive";
  confidence: number | null;
  score_a_heatmap_png: string;
  top_substructures_a: TopSubstructure[];
  score_b_heatmap_png: string;
  top_substructures_b: TopSubstructure[];
}

File: backend/app/schemas/prediction.py
"""
Request/response schemas for the activity prediction endpoints.
"""

from typing import Optional

from pydantic import BaseModel


class PredictionRequest(BaseModel):
    smiles: str
    model_name: str = "Logistic Regression"


class PredictionResponse(BaseModel):
    smiles: str
    model_name: str
    prediction: str
    prediction_label: int
    probability: float
    threshold: float


class ModelInfo(BaseModel):
    name: str
    accuracy: float
    roc_auc: float
    threshold: float


class AvailableModelsResponse(BaseModel):
    models: list[ModelInfo]
    default_model: str


class TopSubstructure(BaseModel):
    token: str
    description: str
    score: float
    percentage: float
    occurrences: int
    direction: str


class HeatmapResponse(BaseModel):
    smiles: str
    model_name: str
    prediction: str
    confidence: Optional[float]
    score_a_heatmap_png: str
    top_substructures_a: list[TopSubstructure]
    score_b_heatmap_png: str
    top_substructures_b: list[TopSubstructure]
